### ISSUE TITLE
Question: Template-fit method with Coffea in Fake Photon analysis

### ISSUE THREAD

**Issue by ico1036** [alexander-held/template_fit_toy#1]
Dear Alexander

First, thank you very much to make a such nice material. 

I'm trying to do template-fit using the output histograms from Coffea to study the fake photon analysis.

There are templates: Data, Real and Fake templates.

The histograms are the sigma-ietaieta distribution with different PT and Eta bins.
( 4 PT-bins and 4 Eta-bins = 16 different histograms per templates )

The purpose is to fit Fake and Real bin-contents to data bin-contents.

ex)
Fake(ith bin) * w1  + Real(ith bin) * w2  -- > Match -->  Data(ith bin)
:Optimize w1 and w2

I have 48 histograms ( three templates * 16 bins )  with the "coffea hist" file.
Also, I have source code for extracting bin,bin-content and data from the Coffea hist file
(https://github.com/JW-corp/J.W_Analysis/blob/main/Coffea_WZG/Jupyter/Data_extractor.ipynb)

I found 2 methods to do this kind of fit ( binned-likelihood fit? ).
: pyhf and cabinetry 

But I don't know which way best fits my purpose.
Could you please recommend the guideline?

Thank you in advance.

**Comment by alexander-held** [comment]
Hi! `cabinetry` makes use of `pyhf`, and adds features and interfaces to make it easier to construct fit models and to interpret fit results. You can also use `pyhf` by itself, especially when you already have histograms produced. `cabinetry` has two main parts:
- creating a workspace (fit model) from columnar inputs, steered by a declarative configuration,
- utilities to perform fits and visualize results, using a `pyhf` workspace as an input.

In your case, the histograms that would be written into the workspace already exist. I have outlined some ideas for this workflow in https://github.com/alexander-held/cabinetry/issues/219, but at the moment it is not supported to the same degree as columnar inputs (ntuples). In your setup, do you want to have one normalization per template and pt/eta region, or do you want to have independent normalization factors per bin? If you want per-bin factors, you would need to make use of [shapefactors](https://pyhf.readthedocs.io/en/v0.6.1/likelihood.html#data-driven-shape-shapefactor) which have not yet been included in `cabinetry`, see https://github.com/alexander-held/cabinetry/issues/212. They should not be difficult to include, but just have not been a priority yet. I am assuming you want the shapefactor scenario, but let me know if I misunderstood the setup.

For the moment, it may be easiest to use `pyhf` directly. Depending on the complexity of your analysis (e.g. inclusion of many systematic uncertainties), it may be easier to use `cabinetry` or [CMS Combine](https://cms-analysis.github.io/HiggsAnalysis-CombinedLimit/) / similar tools (I believe you are in CMS).

To get started with `pyhf`, here is an example for the kind of fit you describe:
```python
import pyhf


def get_parameter_names(model):
    labels = []
    for parname in model.config.par_order:
        for i_par in range(model.config.param_set(parname).n_parameters):
            labels.append(
                f"{parname}[bin_{i_par}]"
                if model.config.param_set(parname).n_parameters > 1
                else parname
            )
    return labels


spec = {
    "channels": [
        {
            "name": "region_one",
            "samples": [
                {
                    "data": [200, 180, 200],
                    "modifiers": [
                        {"name": "scale_real", "type": "shapefactor", "data": None},
                        {"name": "dummy", "type": "normfactor", "data": None},
                    ],
                    "name": "real",
                },
                {
                    "data": [20, 20, 20],
                    "modifiers": [
                        {"name": "scale_fake", "type": "shapefactor", "data": None}
                    ],
                    "name": "fake",
                },
            ],
        },
        {
            "name": "region_two",
            "samples": [
                {
                    "data": [210, 170, 210],
                    "modifiers": [
                        {"name": "scale_real", "type": "shapefactor", "data": None},
                    ],
                    "name": "real",
                },
                {
                    "data": [25, 25, 25],
                    "modifiers": [
                        {"name": "scale_fake", "type": "shapefactor", "data": None}
                    ],
                    "name": "fake",
                },
            ],
        },
    ],
    "measurements": [
        {
            "config": {
                "parameters": [{"name": "dummy", "fixed": True}],
                "poi": "dummy",
            },
            "name": "shapefactor example",
        }
    ],
    "observations": [
        {"data": [250, 210, 230], "name": "region_one"},
        {"data": [265, 215, 245], "name": "region_two"},
    ],
    "version": "1.0.0",
}

ws = pyhf.Workspace(spec)
model = ws.model()
data = ws.data(model)
fit_results = pyhf.infer.mle.fit(data, model)
for i, label in enumerate(get_parameter_names(model)):
    print(f"{label}: {fit_results[i]}")
```

This is using two channels, since with a single channel the model would be over-parameterized (3 bins and 6 free parameters). The factors are correlated between the channels, the same three factors scale the real contribution in both channels (similarly for the fake contribution). The use of the `dummy` parameter is only a workaround, since `pyhf` cannot handle a shapefactor modifier as parameter of interest at the moment. The parameter is set to constant anyway, so even though it shows up in the results, it does not actually do anything here.

**Comment by ico1036** [comment]
Thank you very much @alexander-held  it [works well!](https://github.com/JW-corp/J.W_Analysis/blob/main/Coffea_WZG/Jupyter/Pyhf_example.ipynb)

I have two more questions, sorry for bothering you ( maybe these questions can move to the pyhf github ).

1) Can I select the specific Loss function and Optimizer here? 
I understand that the "pyhf" can use the ML libraries such as pyTorch and TensorFlow.
So, I want to try the Cross-Entropy loss and Gradient Descent optimizer and compare the result.
But I saw the loss(objective) function in MLE is fixed as  [-2LL](https://pyhf.readthedocs.io/en/v0.6.1/_generated/pyhf.infer.mle.twice_nll.html?highlight=mle) in the document.  
In my case, however, I may not need to use a high-performance optimizer because the number of free parameters is small. 


2) What is the best way to see the results of fitting?
I want to check the fit results and also, compare the fit results between "pyhf" and conventional method ex "RooFit".  
I found the method in [document](https://pyhf.readthedocs.io/en/v0.6.1/_generated/pyhf.infer.mle.fit.html)
```python
bestfit_pars, twice_nll = pyhf.infer.mle.fit(data, model, return_fitted_val=True)
twice_nll
```
is this the correct method?


Thank you very much for your help.

**Comment by matthewfeickert** [comment]
> 1. Can I select the specific Loss function and Optimizer here?
>    I understand that the "pyhf" can use the ML libraries such as pyTorch and TensorFlow.
>    So, I want to try the Cross-Entropy loss and Gradient Descent optimizer and compare the result.
>    But I saw the loss(objective) function in MLE is fixed as  [-2LL](https://pyhf.readthedocs.io/en/v0.6.1/_generated/pyhf.infer.mle.twice_nll.html?highlight=mle) in the document.
>    In my case, however, I may not need to use a high-performance optimizer because the number of free parameters is small.

No, `pyhf` is an implementation of the HistFactory statistical model and not meant to be an "open world" of statistical modelling ([`zfit`](https://github.com/zfit/zfit) is more trying to do this like `RooFit` does). So custom objective functions to minimize are not supported, as the use cases that are of interest for HistFactory are those covered in the `pyhf.infer` docs and https://arxiv.org/abs/1007.1727.

> 2. What is the best way to see the results of fitting?
>    I want to check the fit results and also, compare the fit results between "pyhf" and conventional method ex "RooFit".
>    I found the method in [document)[https://pyhf.readthedocs.io/en/v0.6.1/_generated/pyhf.infer.mle.fit.html]
> 
> ```python
> bestfit_pars, twice_nll = pyhf.infer.mle.fit(data, model, return_fitted_val=True)
> twice_nll
> ```
> 
> is this the correct method?

What is your definition of "fit results"? As noted in the `pyhf` docs that you link to (for `v0.6.1`) the return value of [`pyhf.infer.mle.fit`](https://pyhf.readthedocs.io/en/v0.6.1/_generated/pyhf.infer.mle.fit.html) is going to be the return value of the optimizer that you've selected. If your "fit results" are the fit model parameters then those will always be returned, but, as you've noted, if the "fit results" is the objective function evaluated at those best fit model parameters then you need to ask for that with the relevant kwarg. You probably just care about the bestfit model parameters and the relevant test statistic.

---

Aside: @ico1036 I see you're on CMS, so maybe you have thoughts or opinions that you can contribute to https://github.com/scikit-hep/pyhf/issues/344

**Comment by alexander-held** [comment]
@ico1036 while the loss function is fixed, you can change the minimizer. My recommendation is MINUIT (see [the `iminuit` documentation](https://iminuit.readthedocs.io/en/stable/about.html))
```python
pyhf.set_backend("numpy", "minuit")
```
but you can also use `scipy` (default). The MINUIT algorithm has been used for decades and is suitable for HEP use cases. It will also provide you with nuisance parameter uncertainties, which you may be interested in.

Regarding your second question, it depends a bit on what exactly you are looking for. I would suggest having a look at this [`cabinetry` tutorial](https://github.com/cabinetry/cabinetry-tutorials), you can click the ["launch binder button"](https://mybinder.org/v2/gh/cabinetry/cabinetry-tutorials/master?filepath=example.ipynb) to run it interactively. It shows you various visualizations of fit results with `cabinetry`.

For comparison of results with `RooFit`, you may also be interested in [`pyhf-validation`](https://github.com/pyhf/pyhf-validation/).

**Comment by ico1036** [comment]
@matthewfeickert  Thanks a lot!
I means how to check the goodeness of fit.

Also, can I calculate the one weight-factor per template ( not bin )? 
-> W(fake) and W(real) per templatr

**Comment by alexander-held** [comment]
For goodness-of-fit, [`cabinetry` has a goodness-of-fit option](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.fit.fit) which is based on the [saturated model](http://www.physics.ucla.edu/~cousins/stats/cousins_saturated.pdf). You can also visually inspect the fit results, see the `cabinetry` tutorial linked in a previous comment for examples.

If you want one factor to scale a template across bins of a channel (instead of different factors per bin), use a [`normfactor` modifier](https://pyhf.readthedocs.io/en/v0.6.1/likelihood.html#unconstrained-normalisation-normfactor) instead of a `shapefactor` modifier.

**Comment by ico1036** [comment]
Thank you @alexander-held !
I tried the "normfactor" modifier and got a plot. However, I cannot understand the result.
Could you please give me some comments about this result?

![image](https://user-images.githubusercontent.com/13959858/117653299-2dd1c180-b1cf-11eb-9e09-67a13b013086.png)
![image](https://user-images.githubusercontent.com/13959858/117653310-30ccb200-b1cf-11eb-9fd3-dddbef6e74da.png)

Bin with the largest number of the event ( 3rd bin ) shows the worst goodness-of-fit.
I think the Likelihood of this bin is the largest so, this bin should show the best fit. However, bins with low events show good results.

Did I make some mistake on the source code?
https://github.com/JW-corp/J.W_Analysis/blob/main/Coffea_WZG/Jupyter/Pyhf_example.ipynb

or did I misunderstand the fitting process?

Thank you for always helping me.

**Comment by alexander-held** [comment]
The likelihood you are maximizing in the fit is joint across all bins. A better fit to the data in the third bin in your model implies worse fits to the other bins, which explains the result. You have two degrees of freedom (normalization factors) and 7 bins, which will in general mean you cannot get a perfect fit. If the result looks unsatisfactory, it is likely that the model you are fitting is not suitable for the data you are trying to fit. You may need to change to a different model, and possibly consider additional sources of (systematic) uncertainty.
